docs/puzzles/day15.md

```diff
@@ -1,11 +1,98 @@
 import Solver from "../../../../website/src/components/Solver.js"
 
 # Day 15
+By @anatoliykmetyuk
 
 ## Puzzle description
 
 https://adventofcode.com/2021/day/15
 
+## Problem
+The problem in its essence is that of finding the least-costly path through a graph. This problem is solved by Dijkstra Algorithm, nicely explained in this [Computerphile video](https://www.youtube.com/watch?v=GazC3A4OQTE).
+
+## Domain Model
+The two domain entities we are working with are the game map and an individual cell of that map. In presence of the game map, a cell is fully described by a pair of its coordinates.
+
+```scala
+type Coord = (Int, Int)
+```
+
+The game map contains all the cells from the challenge input. It also defines the neighbours of a given cell which we need to know for Dijkstra Algorithm. Finally, it defines a function to get a cost of a given cell.
+
+```scala
+class GameMap(cells: IndexedSeq[IndexedSeq[Int]]):
+  val maxRow = cells.length - 1
+  val maxCol = cells.head.length - 1
+
+  def neighboursOf(c: Coord): List[Coord] =
+    val (row, col) = c
+    val lb = mutable.ListBuffer.empty[Coord]
+    if row < maxRow then lb.append((row+1, col))
+    if row > 0      then lb.append((row-1, col))
+    if col < maxCol then lb.append((row, col+1))
+    if col > 0      then lb.append((row, col-1))
+    lb.toList
+
+  def costOf(c: Coord): Int = c match
+    case (row, col) => cells(row)(col)
+end GameMap
+```
+
+`IndexedSeq` in the `cells` type is important for this algorithm since we are doing a lot of index-based accesses, so we need to use a data structure optimized for that.
+
+## Algorithm – Part 1
+We start the solution by defining three data structures for the algorithm:
+
+```scala
+val visited = mutable.Set.empty[Coord]
+val dist = mutable.Map[Coord, Int]((0, 0) -> 0)
+val queue = java.util.PriorityQueue[Coord](Ordering.by(dist))
+queue.add((0, 0))
+```
+
+The first one is a `Set` of all visited nodes – the ones the algorithm will not look again at. The second one is a `Map` of distances containing the smallest currently known distance from the top-left corner of the map to the given cell. Finally, the third one is a `java.util.PriorityQueue` that defines in which order to examine cells. We are using Java's `PriorityQueue`, not the Scala's one since the Java `PriorityQueue` implementation defines the `remove` operation on the queue which is necessary for efficient implementation and which the Scala queue lacks.
```

Comment:
```suggestion
The first one is a `Set` of all visited nodes – the ones the algorithm will not look at again. The second one is a `Map` of distances containing the smallest currently known distance from the top-left corner of the map to the given cell. Finally, the third one is a `java.util.PriorityQueue` that defines in which order to examine cells. We are using Java's `PriorityQueue`, not the Scala's one since the Java `PriorityQueue` implementation defines the `remove` operation on the queue which is necessary for efficient implementation and which the Scala queue lacks.
```